servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/WriteStreamSubscriber.java

```diff
@@ -307,28 +306,13 @@ void sourceTerminated(@Nullable Throwable cause) {
 
         void channelClosed(Throwable cause) {
             assert eventLoop.inEventLoop();
-            if (hasFlag(CHANNEL_CLOSED) || hasFlag(SUBSCRIBER_TERMINATED)) {
+            if (hasAnyFlags(CHANNEL_CLOSED, SUBSCRIBER_TERMINATED)) {
                 setFlag(CHANNEL_CLOSED);
                 return;
             }
             tryFailure(!written ? new AbortedFirstWrite(cause) : cause);
```

Comment:
If there are pending writes, we should not prematurely terminate the promise but wait for the promise results which will fail if the channel closed before writes were done. The following would be reasonable to do:

```java
void channelClosed(Throwable cause) {
    assert eventLoop.inEventLoop();
    if (hasFlag(CHANNEL_CLOSED)) {
        return;
    }
    if (hasFlag(SUBSCRIBER_TERMINATED)) {
        setFlag(CHANNEL_CLOSED);
        // We have already terminated the subscriber (all writes have finished (one has failed)) then we
        // just close the channel now.
        closeHandler.closeChannelOutbound(channel);
    } else if (activeWrites > 0) {
        // Writes are pending, we will close the channel once writes are done.
        setFlag(CLOSE_OUTBOUND_ON_SUBSCRIBER_TERMINATION);
    } else {
        setFlag(CHANNEL_CLOSED);
        // subscriber has not terminated, no writes are pending and channel has closed so terminate the
        // subscriber with a failure.
        tryFailure(!written ? new AbortedFirstWrite(cause) : cause);
    }
}
```